### PR TITLE
Parse module name stream through modulemd_str

### PR DESCRIPTION
**Pull request by chuazhan**
Use libmodulemd to parse modue name and stream.
One thing to note is no package available in pypi.
Install it through python3-libmodulemd rpm package

JIRA: RHELWF-3779

### CODE REVIEW

**Comment by chuazhan** [comment]
Haven't found a good solution as no libmodulemd package in pypi. Just add some doc in README for developer

**Comment by chuazhan** [comment]
@hluk @qixiang @mvalik Please help to review, thanks!

File: freshmaker/kojiservice.py [hunk 1]
```diff
@@ -407,6 +407,41 @@ def get_bundle_csv(self, build_nvr):
...
+        :param str build_nvr: NVR of module build.
+        :return: a dict for module name and its streams, e.g. {"key": [values]}
+        :rtype: dict
+        """
+        module_name_streams_mapping = dict()
+        build = self.get_build(build_nvr)
+        modulemd_str = build.get("extra", {}).get("typeinfo", {}).get("module", {}).get("modulemd_str")
+        if modulemd_str:
+            try:
+                import gi
```
Comment (qixiang) [inline freshmaker/kojiservice.py:424]: I think there is no strong reason of importing the modules in method instead of the beginning of this module, right?

File: freshmaker/kojiservice.py [hunk 2]
```diff
@@ -407,6 +407,41 @@ def get_bundle_csv(self, build_nvr):
...
+                gi.require_version("Modulemd", "2.0")
+                from gi.repository import Modulemd
+                fedora_index = Modulemd.ModuleIndex.new()
```
Comment (qixiang) [inline freshmaker/kojiservice.py:427]: By checking the API doc, `Modulemd.ModuleIndex` [1] is mainly designed for dealing with repodata, we only need to deal with a modulemd string, so I think using `Modulemd.ModuleStream` [2] could be better.

[1] https://fedora-modularity.github.io/libmodulemd/latest/modulemd-2.0-Modulemd.ModuleIndexMerger.html
[2] https://fedora-modularity.github.io/libmodulemd/latest/ModulemdModuleStream.html

File: freshmaker/kojiservice.py [hunk 3]
```diff
@@ -407,6 +407,41 @@ def get_bundle_csv(self, build_nvr):
...
+                    for module_name in module_names:
+                        module = merged_index.get_module(module_name)
+                        streams = module.get_stream_names()
+                        module_name_streams_mapping[module_name] = streams
+                else:
+                    log.error("Update ModuleIndex failed from %s: %s", modulemd_str, str(failures))
+            except Exception as e:
+                log.error("Parse module name and streams failed from NVR %s: %s", build_nvr, str(e))
+
+        return module_name_streams_mapping
```
Comment (qixiang) [inline freshmaker/kojiservice.py:443]: When error happens, an empty dict is returned, it makes no sense, I'd prefer return `None` if we hit any issue of fetching the module metadata.

Comment (qixiang) [reply freshmaker/kojiservice.py:443]: I believe a module build can only have one stream, not 100% sure, but according to the API names, it should be true, you can get only one stream from a modulemd, there is a `get_stream_name` API, as you see, it's `name` not `names`.

I'd suggest refactoring this API to just return a `Modulemd.ModuleStream` instance, like:
```
def get_modulemd(self, build_nvr):
    ...
    try:
        mmd = Modulemd.ModuleStream.read_string(modulemd_str, False, None, None)
    ...
    # any error happens
    # return None
    return mmd
```
then you can get module name and stream with:
```
mmd = session.get_modulemd("foobar-flatpak-xxxxx")
module_name = mmd.get_module_name()
module_stream = mmd.get_stream_name()
```

File: freshmaker/kojiservice.py (hunk 1 above)
Comment (chuazhan) [reply freshmaker/kojiservice.py:424]: Ack. Probably this is a bad habit from previous experience. We were annoying by cyclic issue, so for an unusual package and it is only used in one place, normally it will be put in lines.

File: freshmaker/kojiservice.py (hunk 2 above)
Comment (chuazhan) [reply freshmaker/kojiservice.py:427]: Ack. Fixing

File: freshmaker/kojiservice.py (hunk 3 above)
Comment (chuazhan) [reply freshmaker/kojiservice.py:443]: Thanks @qixiang for you carefully review and constructive suggestions. I check the doc again, and there is description "modulemd_module_stream_read_string has been deprecated since version 2.11 and should not be used in newly-written code.
Use modulemd_read_packager_string() instead.". I change it with read_packager_string, please have a look.

File: .github/actions/tests/Dockerfile [hunk 4]
```diff
@@ -7,7 +7,9 @@ LABEL \
 
 USER 0
 
-RUN dnf install -y python-devel fedora-packager gcc krb5-devel openldap-devel
+RUN dnf install -y python-devel fedora-packager gcc krb5-devel openldap-devel python3-libmodulemd \
+                   gobject-introspection-devel cairo-devel cairo-gobject-devel
```
Comment (hluk) [inline .github/actions/tests/Dockerfile:11]: Why are these needed? Why not only `python3-libmodulemd`?

File: test-requirements.txt [hunk 5]
```diff
@@ -8,3 +8,4 @@ vcrpy
 requests_mock
 tox
 mypy
+pygobject
```
Comment (hluk) [inline test-requirements.txt:11]: Why is this needed?

File: .github/actions/tests/Dockerfile (hunk 4 above)
Comment (chuazhan) [reply .github/actions/tests/Dockerfile:11]: This is a good question. I was confused too when running the CI. I only install python3-libmodulemd before, but met error like "....Try sudo dnf install gobject-introspection-devel " when installing pygobject

File: test-requirements.txt (hunk 5 above)
Comment (chuazhan) [reply test-requirements.txt:11]: There is no pypi package available for libmodulemd . And actually we are only using gi package. After investigation,  It seems pygobject is good enough. So for testing, we include this package.

File: freshmaker/kojiservice.py [hunk 6]
```diff
@@ -407,6 +409,25 @@ def get_bundle_csv(self, build_nvr):
...
+        """
+        Get Modulemd.ModuleStreamV2 instance through build NVR
+        which can be used to get module's name, stream, etc.
+
+        :param str build_nvr: NVR of module build.
+        :return: Modulemd.ModuleStreamV2 instance.
+        """
+        mmd = None
+        build = self.get_build(build_nvr)
+        modulemd_str = build.get("extra", {}).get("typeinfo", {}).get("module", {}).get("modulemd_str")
```
[the remark this answers is not in the document]
Comment (chuazhan) [reply freshmaker/kojiservice.py:422]: Ok. It is cleaner

File: .github/actions/tests/Dockerfile (hunk 4 above)
Comment (chuazhan) [reply .github/actions/tests/Dockerfile:11]: So I am guessing it is a side-effect for installing pygobject

File: test-requirements.txt (hunk 5 above)
Comment (hluk) [reply test-requirements.txt:11]: I'm confused. Why is this needed only in tests? Is it because of `tox` fails to find the packages installed on the system?

Installing `python3-libmodulemd` should make Modulemd available for Python - I think it should install all necessary packages enable `from gi.repository import Modulemd`.

Comment (chuazhan) [reply test-requirements.txt:11]: Yeah. I had the same idea with you before. I thought that would be ok to install  python3-libmodulemd only. Let me resubmit it again and double check it.

**Comment by chuazhan** [comment]
Sorry guys for the notification spamming. I just want to save some test links for discussion.

File: test-requirements.txt (hunk 5 above)
Comment (chuazhan) [reply test-requirements.txt:11]: `I'm confused. Why is this needed only in tests? Is it because of tox fails to find the packages installed on the system?`
I am guessing the dependencies installed by python3-libmodulemd in system can not be used in tox environment.

Following is the failed cases with its codes:
https://github.com/redhat-exd-rebuilds/freshmaker/runs/4185263908?check_suite_focus=true
https://github.com/redhat-exd-rebuilds/freshmaker/commit/065dfdf2f1d4616d5eb054a17ecf6cead1767010

The error is: ModuleNotFoundError: No module named 'gi'

File: .github/actions/tests/Dockerfile (hunk 4 above)
Comment (qixiang) [reply .github/actions/tests/Dockerfile]: Tests are running inside the tox virtualenvs (without using system site packages), so:

1. First you need `pygobject` to be installed in virtualenvs, that's why you need `pygobject` in `requirements.txt` (it's generated by pip-compile).
2. To allow pip install `pygobject` in virtualenvs, you need some `-devel` packages on system, because they're required to compile `pygobject` and its dependencies (`pycairo` and etc) while installing them within virtualenvs, thus you need to install `gobject-introspection-devel` and `cairo-gobject-devel` with yum/dnf before creating virtualenvs.
3. Fedora systems has `libmodulemd` installed by default, so once you have `pygobject` installed in virtualenvs, it's good enough to import modulemd modules within the virtualenvs. Thus you don't need to install `python3-libmodulemd` with dnf/yum.

**Comment by chuazhan** [comment]
Rebase with the latest codes.

File: README.md [hunk 7]
```diff
@@ -89,6 +89,12 @@ Install the dependencies with:
 python3 setup.py develop
 ```
 
+Install required rpm packages:
+
+```bash
+dnf install -y python-devel fedora-packager gcc krb5-devel openldap-devel gobject-introspection-devel cairo-gobject-devel
```
Comment (hluk) [inline README.md:95]: Would it make sense to drop the package list to `yum-packages-devel.txt` (similar to `yum-packages.txt`)?

```suggestion
sudo dnf install $(sed 's/#.*//' yum-packages-devel.txt)
```

```
❯ cat yum-packages-devel.txt
fedora-packager
gcc
krb5-devel
openldap-devel
python-devel

# pygobject dependencies
cairo-gobject-devel
gobject-introspection-devel
```

File: freshmaker/kojiservice.py [hunk 8]
```diff
@@ -407,6 +412,29 @@ def get_bundle_csv(self, build_nvr):
...
+            modulemd_str = build["extra"]["typeinfo"]["module"]["modulemd_str"]
+        except KeyError:
+            return None
+
+        if modulemd_str:
+            try:
+                _, mmd = Modulemd.read_packager_string(modulemd_str, None, None)
+            except Exception as e:
+                log.error("Read packager string failed for NVR %s: %s", build_nvr, str(e))
+        return mmd
```
Comment (hluk) [inline freshmaker/kojiservice.py:436]:
```suggestion
        try:
            _, mmd = Modulemd.read_packager_string(modulemd_str, None, None)
            return mmd
        except Exception as e:
            log.error("Read packager string failed for NVR %s: %s", build_nvr, str(e))
            return None
```

Comment (chuazhan) [reply freshmaker/kojiservice.py:436]: I know some convention says return immediately, and some says one return point.  I don't have strong opinion to use one not the other.

File: README.md (hunk 7 above)
Comment (chuazhan) [reply README.md:95]: Put it all in a file is ok for me. Just curious about why adding sudo and removing -y?

Comment (hluk) [reply README.md:95]: `sudo` is probably needed for most user but I'm not sure what's the convention for putting it into docs. It's OK to leave it out because it is clear when it is required.

I removed `-y` so that user/dev has chance to review the list of packages before they are installed. I would prefer to have the flag only in scripts.

File: requirements.txt [hunk 9]
```diff
@@ -1,5 +1,5 @@
 #
-# This file is autogenerated by pip-compile with python 3.7
+# This file is autogenerated by pip-compile with python 3.8
```
[the remark this answers is not in the document]
Comment (chuazhan) [reply requirements.txt:2]: yeah. I do refer the DockerFile's fedora version for generating this file, but I am also wondering if the requirements are really dependent on it. If the DockerFile updates the fedora version, we need to update this file as well? I am not sure.

File: README.md (hunk 7 above)
Comment (chuazhan) [reply README.md:95]: > I removed -y so that user/dev has chance to review the list of packages before they are installed. I would prefer to have the flag only in scripts.
Ok. Sounds reasonable to me.

File: requirements.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (chuazhan) [reply requirements.txt:2]: I regenerate this file with fedora34's image and the only difference is the comment
"This file is autogenerated by pip-compile with python 3.9". I think we don't need to
wait for #119

[the remark this answers is not in the document]
Comment (chuazhan) [reply requirements.txt]: Updated

File: README.md (hunk 7 above)
Comment (chuazhan) [reply README.md]: Updated

File: freshmaker/kojiservice.py (hunk 8 above)
Comment (chuazhan) [reply freshmaker/kojiservice.py]: Updated

File: freshmaker/kojiservice.py [hunk 10]
```diff
@@ -407,6 +412,30 @@ def get_bundle_csv(self, build_nvr):
...
+        :return: Modulemd.ModuleStreamV2 instance.
+        """
+        mmd = None
+        build = self.get_build(build_nvr)
+        try:
+            modulemd_str = build["extra"]["typeinfo"]["module"]["modulemd_str"]
+        except KeyError:
+            return None
+
+        if modulemd_str:
```
Comment (hluk) [inline freshmaker/kojiservice.py:431]: This should not be needed now.

**Review by hluk — APPROVED** [review]

File: freshmaker/kojiservice.py (hunk 10 above)
Comment (chuazhan) [reply freshmaker/kojiservice.py:431]: Ah. Yes, Empty string should be covered in following codes.

**Comment by chuazhan** [comment]
@gnaponie Can you help to have a review, thanks!(Sorry for not involving you at the beginning of task. Just in the process of being familiar with the team and project)

File: tests/test_kojiservice.py [hunk 11]
```diff
@@ -142,3 +142,19 @@ def test_get_bundle_csv_unavailable(mock_get, mock_koji, mock_log):
...
...\n      rpms:\n      - ghc\n    minimal:\n      description: just compiler and base\n      rpms:\n      - ghc-base-devel\n    small:\n      description: compiler with main core libs\n      rpms:\n      - ghc-devel\n  components:\n    rpms:\n      ghc:\n        rationale: compiler\n        repository: git+https://src.fedoraproject.org/rpms/ghc\n        cache: https://src.fedoraproject.org/repo/pkgs/ghc\n        ref: 9.2\n        buildorder: 1\n        arches: [aarch64, armv7hl, i686, ppc64le, s390x, x86_64]\n...\n', 'name': 'ghc', 'stream': '9.2', 'module_build_service_id': 13274, 'version': '3620211101111632', 'context': 'd099bf28', 'content_koji_tag': 'module-ghc-9.2-3620211101111632-d099bf28'}}}, 'id': 1850907, 'name': 'ghc', 'nvr': 'ghc-9.2-3620211101111632.d099bf28', 'owner_id': 56, 'owner_name': 'petersen', 'package_id': 1853, 'package_name': 'ghc', 'release': '3620211101111632.d099bf28', 'source': 'https://src.fedoraproject.org/modules/ghc.git?#5405c8a084c04476a956f1b3e49a733b888066a5', 'start_time': '2021-11-01 13:48:17+00:00', 'start_ts': 1635774497.0, 'state': 1, 'task_id': None, 'version': '9.2', 'volume_id': 0, 'volume_name': 'DEFAULT', 'cg_name': 'module-build-service'}
```
Comment (gnaponie) [inline tests/test_kojiservice.py:150]: Could you put this on multiple lines for readability? Also you can remove fields that you don't need for the test, since this is a test.

File: freshmaker/kojiservice.py [hunk 12]
```diff
@@ -407,6 +412,29 @@ def get_bundle_csv(self, build_nvr):
...
+        """
+        mmd = None
+        build = self.get_build(build_nvr)
+        try:
+            modulemd_str = build["extra"]["typeinfo"]["module"]["modulemd_str"]
+        except KeyError:
+            return None
+
+        try:
+            _, mmd = Modulemd.read_packager_string(modulemd_str, None, None)
```
Comment (gnaponie) [inline freshmaker/kojiservice.py:432]: I'm not familiar with Modulemd, could you explain it a bit to me? Thanks.

**Comment by gnaponie** [comment]
> @gnaponie Can you help to have a review, thanks!(Sorry for not involving you at the beginning of task. Just in the process of being familiar with the team and project)

No worries! :)

**Comment by gnaponie** [comment]
Seems like you have a conflict on the freshmaker/kojiservice.py file. Could you resolve it?

File: freshmaker/kojiservice.py (hunk 12 above)
Comment (chuazhan) [reply freshmaker/kojiservice.py:432]: This is a library to manipulate module metadata files. More detail, you can check https://github.com/fedora-modularity/libmodulemd
I am using this api https://fedora-modularity.github.io/libmodulemd/latest/modulemd-2.0-Modulemd.html#modulemd-read-packager-string   to parse the module's name and stream from a yaml string.

File: tests/test_kojiservice.py (hunk 11 above)
Comment (chuazhan) [reply tests/test_kojiservice.py:150]: I was thinking to save the actual data and maybe in future we can reference different attributes other than name and stream.
But sure, let's make it more readable.

**Comment by chuazhan** [comment]
> Seems like you have a conflict on the freshmaker/kojiservice.py file. Could you resolve it?

Sure. Rebased and upated

File: freshmaker/kojiservice.py [hunk 13]
```diff
@@ -37,11 +37,16 @@
...
 import requests
 import yaml
-import freshmaker.utils
-from freshmaker import log, conf, db
-from freshmaker.consumer import work_queue_put
-from freshmaker.events import BrewContainerTaskStateChangeEvent
-from freshmaker.models import ArtifactBuild
+
+import gi
+gi.require_version('Modulemd', '2.0')  # noqa E402
```
Comment (hluk) [inline freshmaker/kojiservice.py:42]: Any `import` that goes under this, needs `# noqa E402`. :( Can we perhaps hide this under separate module (e.g. `freshmaker.modulemd`)?

File: tests/test_kojiservice.py [hunk 14]
```diff
@@ -142,3 +142,37 @@ def test_get_bundle_csv_unavailable(mock_get, mock_koji, mock_log):
...
...tps://src.fedoraproject.org/modules/ghc.git?#5405c8a084c04476a956f1b3e49a733b888066a5\n      ursine_rpms:\n      - ghc-bytestring-devel-0:0.10.12.0-115.fc35.s390x\n      - ghc-exceptions-prof-0:0.10.4-115.fc35.armv7hl\n  dependencies:\n  - buildrequires:\n      ghc: [8.10]\n      platform: [f36]\n    requires:\n      platform: [f36]\n  references:\n    community: https://www.haskell.org/ghc\n    documentation: https://wiki.haskell.org/GHC\n    tracker: https://gitlab.haskell.org/ghc/ghc/issues\n  profiles:\n    all:\n      description: standard installation\n      rpms:\n      - ghc\n      - ghc-doc\n      - ghc-prof\n    default:\n      description: standard installation\n      rpms:\n      - ghc\n    minimal:\n      description: just compiler and base\n      rpms:\n      - ghc-base-devel\n    small:\n      description: compiler with main core libs\n      rpms:\n      - ghc-devel\n  components:\n    rpms:\n      ghc:\n        rationale: compiler\n        repository: git+https://src.fedoraproject.org/rpms/ghc\n        cache: https://src.fedoraproject.org/repo/pkgs/ghc\n        ref: 9.2\n        buildorder: 1\n        arches: [aarch64, armv7hl, i686, ppc64le, s390x, x86_64]\n...\n',
```
Comment (gnaponie) [inline tests/test_kojiservice.py:152]: Is this needed to be that long for this test? I'm concerned (besides readability) that flake8 will fail.

File: freshmaker/kojiservice.py (hunk 12 above)
Comment (gnaponie) [reply freshmaker/kojiservice.py:455]: Thank you!

**Comment by gnaponie** [comment]
Anyhow, besides the very small comment (you're free to consider that as optional) the change looks good to me. +1

**Review by gnaponie — APPROVED** [review]

**Comment by chuazhan** [comment]
> Anyhow, besides the very small comment (you're free to consider that as optional) the change looks good to me. +1 I deeply apologize for the delay in the review, I had some days of PTO. Thank you for the patience!

Thanks for the review. No problem at all.

File: tests/test_kojiservice.py (hunk 14 above)
Comment (chuazhan) [reply tests/test_kojiservice.py]: Ack. Make it shorter

File: freshmaker/kojiservice.py (hunk 13 above)
Comment (chuazhan) [reply freshmaker/kojiservice.py]: Yes. But it seems sth come be better, change it a bit. 
I think this is a package we will use, not wrapping a new service, so I prefer not to put it in a separate module.

I tried to put import gi, gi.require_version('Modulemd', '2.0')  # noqa E402 at the end of import chain and there is no flake8 issue, but I guess there is some convention to put the freshmaker model at the end.

**Comment by chuazhan** [comment]
Rebased and updated

File: freshmaker/kojiservice.py (hunk 13 above)
Comment (hluk) [reply freshmaker/kojiservice.py]: The idea was to create simple file `freshmaker/modulemd.py`:

```python
import gi
gi.require_version('Modulemd', '2.0')
from gi.repository import Modulemd  # noqa E402
```

And use `from freshmaker.modulemd import Modulemd`.

Comment (chuazhan) [reply freshmaker/kojiservice.py]: Yeah. I agree that this will resolve the flake issues. But do we have strong opinion to create a new file for fixing flake8 issues, is this a convention for fixing this kind of issues, I am not sure.
